packages/tasit-action/src/TasitAction.js

```diff
@@ -19,117 +22,370 @@ class Utils {
+
+  removeAllListeners = () => {
+    this.#events.forEach(event => {
+      this.removeListener(event.eventName, event.listener);
+    });
+  };
+
+  unsubscribe = () => {
+    this.removeAllListeners();
+  };
+
+  // TODO: Make protected
+  emitErrorEvent = error => {
+    this.#events.forEach(event => {
+      if (event.eventName === "error") {
+        event.listener(error);
+      }
+    });
+  };
+
+  // TODO: Make protected
+  addErrorListener = listener => {
+    const eventName = "error";
+    this.#events.push({
+      eventName,
+      wrappedEventName: eventName,
+      listener,
+      wrappedListener: listener,
+    });
+  };
+
+  // TODO: Make protected
+  getEventInfo = (wrappedEventName, wrappedListener) => {
+    return this.#events.find(
+      event =>
+        event.wrappedEventName === wrappedEventName &&
+        event.wrappedListener === wrappedListener
+    ).info;
+  };
+
+  // TODO: Make protected
+  setEventInfo = (wrappedEventName, wrappedListener, info) => {
+    this.#events.forEach(event => {
+      if (
+        event.wrappedEventName === wrappedEventName &&
+        event.wrappedListener === wrappedListener
+      )
+        event.info = info;
+    });
+  };
+
+  // TODO: Make protected
+  addListener = (eventName, wrappedEventName, listener, wrappedListener) => {
+    this.#emitter.on(wrappedEventName, wrappedListener);
+
+    this.#events.push({
+      eventName,
+      wrappedEventName,
+      listener,
+      wrappedListener,
+    });
+
+    setTimeout(() => {
+      this.emitErrorEvent(new Error(`Listener removed after reached timeout`));
+      this.removeListener(eventName, listener);
+    }, EVENTS_CONFIG.timeout);
+  };
+}
+
+class TransactionSubscription extends Subscription {
   #txPromise;
   #provider;
+  #tx;
 
   constructor(txPromise, provider) {
+    super(provider);
     this.#txPromise = txPromise;
     this.#provider = provider;
   }
 
-  on = async (trigger, callback) => {
-    if (trigger !== "confirmation") {
-      throw new Error(`Invalid subscription trigger, use: ["confirmation"]`);
+  on = async (eventName, listener) => {
+    const triggers = ["confirmation", "error"];
+
+    if (!triggers.includes(eventName)) {
+      throw new Error(`Invalid listener trigger, use: [${triggers}]`);
     }
 
-    if (!callback || typeof callback !== "function") {
-      throw new Error(`Cannot subscribe without a function`);
+    if (!listener || typeof listener !== "function") {
+      throw new Error(`Cannot listening without a function`);
     }
 
-    const blockHeight = 1;
-    const tx = await this.#txPromise;
+    if (eventName === "error") {
+      this.addErrorListener(listener);
+      return;
+    }
 
-    this.#provider.on(tx.hash, async receipt => {
-      const { confirmations } = receipt;
-      const message = {
-        data: {
-          confirmations: confirmations,
-        },
-      };
+    this.#tx = await this.#txPromise;
 
+    let wrappedListener = async blockNumber => {
```

Comment:
Abstracting, there are two kinds of events `error` and `non-error`. `Error event` doesn't have a timeout and don't use wrappers. Error listener addition could live inside addListener (with an if). Do you think that will be better?

Reply:
A) Even with no modifications to the code, a function called addConfirmationListener still seems warranted to make this block of code smaller.

B) With an abstraction like "wrap if needed" for all listeners, we wouldn't need a separate flow based on the wrapper distinction.

C) Or maybe we wrap errors with a listener that does nothing. We might want to consider an error wrapper if we wanted to intercept specific ethers.js errors and modify their error message to be more user-friendly, for instance.

D) I would argue that ContractEvent shouldn't have a timeout either.

E) Actually, I would argue that even the confirmation listener shouldn't have a timeout. Internally within the SDK, the event for a "confirmation-timeout" (exact name T.B.D.) will only be fired after a timeout since the last confirmation, though. But no listener will be removed automatically after a timeout without an explicit `removeListener`

F) Yeah, I think it makes sense to add all listeners via addListener and conditionals inside possibly calling more specialized smaller functions (with additional params if desired)

Reply:
Oh, thank you for the clarification!
A) Got it. The code is better now.
B) I agree.
C) Ok, it makes sense. Errors could demand wrappers as well.
D) Sure, I've pushed down setTimeout to TransactionSubscription class.
E) It's pending.
F) Ok.

Reply:
A) Cool!
B) Cool.
C) Cool.
D) Great, thanks.
E) What do you mean by "pending"? Do you mean that you'll be working on this soon? If so, cool.
F) Cool.

Reply:
Removing the time-out can be split into a separate GitHub issue that's a "good first issue"

Reply:
Refs: https://github.com/tasitlabs/TasitSDK/issues/96

Reply:
I'm resolving this thread since we have an issue tracking it now

Reply:
@marcelomorgado Or wait - is there anything from A B C D or F that's still untracked before we resolve this?

Reply:
I think not, wrapper/refactoring commented here seems are duplicated with other threads and will be affected (maybe deprecated) when 1-listener change will be made.

Reply:
Okay, sounds good! I'll resolve this.